content/posts/2025-04-15_official-qgis-server-docker-image.md

```diff
@@ -0,0 +1,349 @@
+<ServerException>Project file error. For OWS services: please provide a SERVICE and a MAP parameter pointing to a valid QGIS project file</ServerException>
+```
+
+Here, we observe a `500` error code from QGIS Server indicating that the `fake.qgs` project specified via `REQUEST_URI="MAP=fake.qgs"` does not exist. The exception `<ServerException>Project file error.</ServerException>` is therefore returned by QGIS Server.
+
+!!! note
+    `Xvfb :99` starts a virtual X server with the display number `:99`, meaning that graphical applications running with this X server instance will not be displayed on a real monitor but will be rendered in memory. Once this virtual X server is started, the environment variable `ENV DISPLAY :99` is set in the [Dockerfile](https://github.com/qgis/qgis-docker/blob/main/server/Dockerfile) so that the QGIS Server application knows where to render in memory.
+
+### Startup script
+
+Previously, we started the QGIS Server container in interactive mode using the `-i` option and the `/bin/bash` command. Without these options, the QGIS Server application starts normally based on the `CMD` or `ENTRYPOINT` instruction specified in the Dockerfile.
+
+The startup script used, located in the container's filesystem, can be found at `/usr/local/bin/start-xvfb-nginx.sh`, and its path can be obtained by inspecting the image.
+
+{% raw %}
+
+```bash title="Inspecting the image to locate the startup script"
+# Retrieving the startup script path
+$ docker inspect -f '{{.Config.Cmd}}' qgis/qgis-server:ltr
+[/bin/sh -c /usr/local/bin/start-xvfb-nginx.sh]
+
+# Displaying the contents of the startup script
+$ docker run qgis/qgis-server:ltr cat /usr/local/bin/start-xvfb-nginx.sh
+```
+
+{% endraw %}
+
+By examining the contents of this script, we can observe the startup sequence of the third-party utilities mentioned above:
+
+- the graphical server `Xvfb`
+- `spawn-fcgi`, which launches QGIS Server while specifying TCP port `9993` for communication between the web server and the FCGI process
+- `NGINX`, if needed, depending on the `SKIP_NGINX` environment variable set at runtime
+
+```bash title="Retrieving the path of the startup script"
+[...]
+/usr/bin/Xvfb :99 -ac -screen 0 1280x1024x16 +extension GLX +render -noreset >/dev/null &
+XVFB_PID=$(waitfor /usr/bin/Xvfb)
+
+if [ -z "$SKIP_NGINX" ] || [ "$SKIP_NGINX" == "false" ] || [ "$SKIP_NGINX" == "0" ]; then
+    nginx
+    NGINX_PID=$(waitfor /usr/sbin/nginx)
+fi
+
+spawn-fcgi -n -u ${QGIS_USER:-www-data} -g ${QGIS_USER:-www-data} -d ${HOME:-/var/lib/qgis} -P /run/qgis.pid -p 9993 -- /usr/lib/cgi-bin/qgis_mapserv.fcgi &
+[...]
+```
+
+### NGINX configuration
+
+![logo nginx](https://cdn.geotribu.fr/img/logos-icones/logiciels_librairies/nginx.svg){: .img-thumbnail-left }
+
+The NGINX web server configuration is deployed by replacing the default configuration file `/etc/nginx/nginx.conf`:
+
+```bash title="Displaying the contents of the NGINX configuration file"
```

Comment:
render is ugly. Could you add some words to make the block title look nicer?

![image](https://github.com/user-attachments/assets/898b4514-b951-4c3a-ad18-62b3ad609a5d)

Reply:
I pushed a more verbose version :crossed_fingers:

Reply:
Well done, it does the trick.